### DISCUSSION TITLE
[ecdk] Looking for unknown solution

### DISCUSSION THREAD

**Discussion by kkafar** [kkafar/master-monorepo#295] [Ideas]
# What

Basically I do have access to files with solutions for each instance and we can verify whether a solutions returned from solver are already known, or maybe we have found a new one.

Each instance has it's own solution file. Inside this file, each line is a solution string as described in problem model with single difference: operation numbers are separated with white spaces (single space) instead of underscores (`_`).

# How

I've already issue with high RAM usage due to loading whole experiment raw data all at once in analyze phase, so as these solution files can be quite big (tens of thousands of solution known for few instances) I do not want to load them into memory.

> [!note]
TODO: We can change the processing flow a bit to compute it experiment by experiment allowing us to load data only from single experiment at the same time. After processing each experiment we would memorize only the aggregate data so that we can compute final statistics.

Simultaneously I do need to query this data for every unique solution string I've found, so assuming that I do not load everything at once, it would require multiple filesystem / file opening operations & data processing (changing ` ` -> `_` and computing md5 hashes).

Hence I think it would be ok to create script to preprocess this data once, compute hashes for each solution string & put it in sqlite database which I can later query.

# Implementation plan details

The script will be fairly easy to implement. One thing to discuss is exact database schema.

## DB details

Basically the db should live in `ecdk` directory maybe in `ecdk/data/solutions.db` and should not be put under VCS.

The schema should be simple:

`SolutionReference` table:

```csv
experiment_id | solution_hash | full_solution
```

`NewSolutions` table:

```csv
batch_id | experiment_id | series_id | solution_hash | full_solution
```

All of these are just strings (`TEXT`, `UTF-8` encoded).

## Flow

This (new solution finding) should be done in `analyze` process after solution validation. Due to how this is implemented right now it might be easier to do this before plotting but I do not see any dependency here, so this can be decided on an interim basis during implementation.

## DB existence assurance

The `ecdk` script should assure that the db exists and create it if needed.
